### ISSUE TITLE
Basic credentials not used when adding a GeoJSON https vector

### ISSUE THREAD

**Issue by MichaelBitard** [qgis/QGIS#49559]
### What is the bug or the crash?

When I try to add an HTTPS GeoJSON vector with basic auth, I see, by debugging the server, that it receives on authenticated call (ie. one with the `Authorization` header set) followed immediately by 3 non authenticated calls.

### Steps to reproduce the issue

1. Go to Laye
2. Click on Data Source Manager
3. Click on Vector in the Popup
4. Select the `Protocol: https, cloud, etc` source type radio button
5. Select the GeoJSON type in the Protocol select
6. Type an URI (A server that responds differently with and without authentication)
7. Add a Basic authentication
8. See that the results you get are the ones without authentication

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.24.3-Tisler | QGIS code branch | Release 3.24
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.4
GDAL/OGR version | 3.4.3
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.36.0
PDAL version | 2.4.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.5
QScintilla2 version | 2.13.0
OS version | Fedora Linux 36 (Workstation Edition)
  |   |   |  

</body></html>QGIS version
3.24.3-Tisler
QGIS code branch
[Release 3.24](https://github.com/qgis/QGIS/tree/release-3_24)
Qt version
5.15.3
Python version
3.10.4
GDAL/OGR version
3.4.3
PROJ version
8.2.1
EPSG Registry database version
v10.041 (2021-12-03)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.36.0
PDAL version
2.4.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.5
QScintilla2 version
2.13.0
OS version
Fedora Linux 36 (Workstation Edition)

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by nicogodet** [comment]
3.24.3 is EOL, can you try with Master or 3.26.1 or 3.22.9 ?

**Comment by MichaelBitard** [comment]
The bug is present on 3.22.5. I cannot test on 3.24.3 nor 3.22.9 on fedora, it's not present in the copr repositories yet.

I'll try to build from master

**Comment by vmaubert** [comment]
Hi, I tried with [3.26.1-Buenos Aires](https://github.com/qgis/QGIS/commit/b609df9ed4) version. The bug is present. :-/

**Comment by gioman** [comment]
@MichaelBitard what is the real issue here, that the layer is not loaded? To test can I use a geojson behind apache basic http authentication for example?

**Comment by MichaelBitard** [comment]
@gioman yes, we see in our debug session that 4 api calls are triggered by QGis, the first one is correctly authenticated (we see the Authorization header) and the next 3 does not have the Authorization header set.
So I think if you do what you said, you should see the bug indeed.

If not, I could try and setup a small python server to reproduce the problem

**Comment by gioman** [comment]
> So I think if you do what you said, you should see the bug indeed.

@MichaelBitard I placed a sample geojson file under apache control and added basic apache http auth, and I was able to use the credentials to load such file in QGIS.

**Comment by MichaelBitard** [comment]
@gioman Ok I think we narrowed the issue here. It is still a QGis bug, a little bit more tricky than I first thought.

Our server does **not** forbid an unauthenticated request, it responds with fewer results.

I wrote a small python program (you "just" need to have flask installed), that you can launch and try.
You can put any credential, it does not mind.

We have 2 routes, 
1- `http://localhost:5000/test` which will respond an empty geojson when no basic creds are set, and the correct geojson when creds are present
2- `http://localhost:5000/private` which will behave like your apache test (allow/forbid)


By testing 1 you see that there is no geojson displayed in qgis, and in the python logs, a bunch of 401

By testing 2 you see the valid geojson displayed in qgis, BUT a lot of 401 immediately followed by 200 in the python logs

IMHO, the 401 errors in the logs are the actual bug in QGIS, it should *never* try a non authenticated request when you specifically set auth.

Bonus question: Why are there 4 requests being done?

The python code: 

```python
from flask import Flask, jsonify, request, Response

app = Flask(__name__)

geo_json = {
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "geometry": {
                "type": "MultiPolygon",
                "coordinates": [
                    [[[-53.8654963071556, 5.085232839853103], [-53.867199047259476, 5.081050258018496], [-53.85051438993294, 5.07422038294979], [-53.848811501516764, 5.078402972071207], [-53.8654963071556, 5.085232839853103]]]
                ]
            },
            "properties": {}
        }
    ]
}

@app.route('/test', methods=['GET'])
def test():
    auth_header = request.headers.get('Authorization')
    if auth_header is None:
        print('we do not have an auth header')
        return jsonify({"type": "FeatureCollection", "features": []})
    print('we have an auth header', auth_header)
    return jsonify(geo_json)

@app.route('/private', methods=['GET'])
def private():
    auth_header = request.headers.get('Authorization')
    if auth_header is None:
        resp = Response("Authentication Required")
        resp.headers['WWW-Authenticate'] = 'Basic realm="Authentication Required"'
        resp.status_code = 401
        return resp
    return jsonify(geo_json)


if __name__ == '__main__':
    app.run(debug=True)
```

**Comment by gioman** [comment]
> It is still a QGis bug, a little bit more tricky than I first thought.

@MichaelBitard do you get the same if you use OGR? https://gdal.org/user/virtual_file_systems.html

**Comment by MichaelBitard** [comment]
@gioman I'm not sure what you mean by that :)

**Comment by MichaelBitard** [comment]
@gioman I tried some stuff with orginfo (which I'm really not familiar with) and I think I get some correct answers.
`ogrinfo -ro -al -so /vsicurl/http://user:pass@localhost:5000/test` --> return the correct data (and only one request is received by the server)
`ogrinfo -ro -al -so /vsicurl/http://localhost:5000/test` --> returns an empty set (that's expected)

If I had to make a wild guess, I'd say it could be related to the cache system of qgis.

Oh, and we noticed something quite strange (but related to how basic auth rejection works), if, in the python code, we remove the line `resp.headers['WWW-Authenticate'] = 'Basic realm="Authentication Required"'` then the private route does not work on QGis anymore, it does not retries with credentials when it gets rejected by trying without credentials. Maybe it can help pinpoint the issue

**Comment by gioman** [comment]
> and I think I get some correct answers

@MichaelBitard thanks for checking, so it seems really a QGIS issue.
